modules/dnn/src/dnn.cpp

```diff
@@ -1925,7 +1930,14 @@ struct Net::Impl
             // Transfer data to CPU if it's require.
             ld.outputBlobsWrappers[pin.oid]->copyToHost();
         }
-        return ld.outputBlobs[pin.oid];
+
+        if (preferableTarget == DNN_TARGET_OPENCL_FP16)
+        {
+            convertFp16(ld.outputBlobs[pin.oid], output_blob);
```

Comment:
I think it's better to check if `ld.outputBlobs[pin.oid]` contains fp16 values rather `preferableTarget == DNN_TARGET_OPENCL_FP16` because if `ld.outputBlobs[pin.oid]` has fp32 type `output_blob` will has fp16 one.